msal/token_cache.py

```diff
@@ -37,18 +37,20 @@ def __init__(self):
 
     def find(self, credential_type, target=None, query=None):
         target = target or []
+        target = [item.lower() for item in target]
         assert isinstance(target, list), "Invalid parameter type"
         target_set = set(target)
         with self._lock:
             # Since the target inside token cache key is (per schema) unsorted,
             # there is no point to attempt an O(1) key-value search here.
             # So we always do an O(n) in-memory search.
-            return [entry
-                for entry in self._cache.get(credential_type, {}).values()
-                if is_subdict_of(query or {}, entry)
-                and (target_set <= set(entry.get("target", "").split())
-		    if target else True)
-                ]
+            cache_entries = []
+            for entry in self._cache.get(credential_type, {}).values():
+                target_cmp = entry.get("target", "").lower()
+                if is_subdict_of(query or {}, entry) and (target_set <= set(target_cmp.split())
+                if target else True):
```

Comment:
>if target else True): [](start = 16, length = 21)

why is this needed? Will it ever evaluate to false?

Reply:
This is the part of second expression after the and operator. So we evaluate the expression only if target is a non empty list. If target is an empty list we want it to be True and skip evaluating the expression. Target can be an empty list in case of finding a refresh token. So if the first part is True in case of Refresh Token , the cache entry is valid and should be appended.

Reply:
hmm, looks strange to me. will add this to my training list

---
In reply to: [269679161](https://github.com/AzureAD/microsoft-authentication-library-for-python/pull/26#discussion_r269679161) [](ancestors = 269679161)